dwave/plugins/torch/samplers/block_spin_sampler.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from typing import TYPE_CHECKING, Callable, Literal
+
+import torch
+from torch import nn
+
+from dwave.plugins.torch.nn.functional import bit2spin_soft
+from dwave.plugins.torch.models.boltzmann_machine import GraphRestrictedBoltzmannMachine as GRBM
+from dwave.plugins.torch.tensor import rands
+
+__all__ = ["BlockSpinSampler"]
+
+
+class BlockSpinSampler(nn.Module):
+    """A block-spin update sampler for graph-restricted Boltzmann machines.
+
+    Note this is a sampler and not a neural network. It extends `nn.Module` for the convenience of
+    managing devices of indices (stored as parameters).
+
+    Due to the sparse definition of GRBMs, some tedious, and ugly, indexing tricks are required for
+    efficiently sampling in blocks of spins. Ideally, an adjacency list can be used, however,
+    adjacencies are ragged, making vectorization inapplicable.
+
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite
+    temperature which, at fixed parameters, converge upon Boltzmann distributions. Block-Metropolis
+    allows higher acceptance rates for proposals (faster single-step mixing), but is non-ergodic in
+    the limit of zero temperature. Decorrelation from an initial condition can be slower.
+    Block-Gibbs represents best practice for independent sampling.
+
+    Args:
+        grbm (GRBM): The Graph-Restricted Boltzmann Machine to sample from.
+        crayon (Callable): A colouring function; a function that maps a single node of the `grbm` to
```

Comment:
Callable[[int], int], to better specify the function signature.

Reply:
Using Hashable instead